### PR TITLE
feat: add hide / unhide preview or results and prompt

### PR DESCRIPTION
**Pull request by johnybx**
Add new options:
- to hide and unhide preview leaving just results and prompt.
- to hide and unhide results and prompt leaving just preview window active. Active preview buffer inherits autocmd for `BufLeave` and config mappings from prompt which makes it possible to also cycle through previews. 

Example key mapping:
```lua
require("telescope").setup({
	defaults = {
		mappings = {
			n = {
				["<C-h>"] = actions.toggle_preview,
				["<C-j>"] = actions.toggle_results_and_prompt,
			},
			i = {
				["<C-h>"] = actions.toggle_preview,
				["<C-j>"] = actions.toggle_results_and_prompt,
			},
               }
       }
})
```

This still require little tweaks for example I am not really sure how should I keep selection ( or if it is currently possible )  for specific entry when hiding preview in vertical layout because the size of results window changes and picker refresh also change the order of results.

### CODE REVIEW

**Comment by fdschmidt93** [comment]
Many thanks for your PR! :)

At first crack your PR works like a charm in the standard layout; however, it seems to have some issues in

1. dropdown theme:
  * layout is not 1-to-1 consistent with how preview is toggled in standard layout as results are not fully filled
  * middle border line underneath prompt vanishes and layout changes afterwards
2. ivy: breaks more materially, just give it a whirl :)

Another (a bit degenerate) question: how to toggle previewer if the picker was launched with `previewer = false` (or `preview = false` in config)? Fairly enough, doesn't seem to work yet.

I'll try to look at the code some time tomorrow.

**Comment by johnybx** [comment]
Yeah, right now I am testing with layouts which I personally use ti catch bugs which are mostly horizontal and vertical layout but I plan to also check other layouts in the following days. At the moment it felt like quite change so I wanted to have some opinion if I am not going wrong way. Also I would like opinion on [this change](https://github.com/nvim-telescope/telescope.nvim/pull/1305/files#diff-5f349484705c8602e40e19020e0613073598e4b06a1044131df314b8867651e3R391) because I am not really sure how would I go around that :thinking: .

I didn't checked how could I support the case with `previewer = false` yet even though I think that it is something what might people want :thinking:

**Comment by tami5** [comment]
OHHHH SOO LOOKING FORWARD FOR THIS, Thanks

**Comment by fdschmidt93** [comment]
I'll do my very best to dig through this PR exhaustively tomorrow (cannot promise I'm afraid) and latest by Friday :)

> Also I would like opinion on this change because I am not really sure how would I go around that

https://github.com/nvim-telescope/telescope.nvim/pull/797 introduced this change. Do you _have_ to remove `vim.schedule` to make your PR work? What is the error caused here? My hunch is this would just need a `nvim_win_is_valid` in your case on top. Admittedly, I haven't fully grokked yet what exact flickering is supposedly reduced as it doesn't really seem to be making a difference when previewing files from the telescope repo?

**Comment by johnybx** [comment]
The problem with `vim.schedule` at that specific place was that when doing refresh of preview while results and prompt are hidden I need to add mapping to new preview buffer [here](https://github.com/nvim-telescope/telescope.nvim/pull/1305/files#diff-9ac2063a221b713185ddb6df3157cf01edf3b12894c2ae984e14250bf16ebd89R1179) but when `vim.api.nvim_win_set_buf(status.preview_win, bufnr)` is in `vim.schedule` then there is raise condition that schedule is not always executed before the `a.nvim_win_get_buf(status.preview_win)` is executed which means that mapping are apply to old buffer. I will check the related issue later today but as far as I can say this change did no cause anything noticeable for me.

**Comment by fdschmidt93** [comment]
> I didn't checked how could I support the case with previewer = false yet even though I think that it is something what might people want thinking

Yes, this is what users were primarily seeking, cf. #643 (this PR should close that issue).  As means of background, whenever you pass `previewer = false` to a picker, this is a preferred option over the default picker as per the `pickers.new`. So here's one way to make this more or less cleanly:
1. Catch `previewer = false` and redirect accordingly (meaning you instantiate the previewer in the background but just not use it)
2. Manage `picker.all_previewers` (and associated layouts) that you essentially toggle between all previewers when running your action

On 2. I think the original idea of `picker.all_previewers` was a different one, though I'll have to go through your code tomorrow and think about this a bit more carefully

**Comment by johnybx** [comment]
I checked how could I resolve the `previewer = false` case but I don't really like having to search telescope default which are not even currently exported in `config.lua`. I think more user friendly is leave the configuration of previewer possible for user and introduce option `disable_previewer_at_startup` - user can still choose previewer but preview is hidden at first and can be toggled with `actions.toggle_preview` or `actions.toggle_results_and_prompt`.

Example config:
```lua
require("telescope").setup({
	defaults = {
		...
       },
        pickers = {
		buffers = {
			disable_previewer_at_startup = false,
		},
       }
})
```

Is it fine to have this new option ?

**Comment by Conni2461** [comment]
I dont like a new options. Its unnecessary.

Look at this code: https://github.com/nvim-telescope/telescope.nvim/blob/ec487779909ebc8bffbbe054a2e8a4fbe05c0e51/lua/telescope/pickers.lua#L1126-L1152

Left hand side is `opts.previewer = false`. Right hand side is `defaults.previewer = fn` (or table of fn). Rather than just merging both and prioritizing opts, take `defaults.previewer` out and hide it in results with `__`.

**Comment by l-kershaw** [comment]
I'm away from my PC this week, so am not able to have a proper look through this PR at the moment. Apologies if I misunderstand anything because of this 😅

---

I agree that toggling the preview is often useful, but like @Conni2461 I think that toggling results and prompt would likely be more trouble than it's worth in Telescope, as it is primarily built to be a fuzzy finder, and is heavily centered around the prompt and results. Removing them would introduce the potential for lots of issues. If the purpose of doing this is to maximise the size of the preview, then I think this can be done sufficiently in other ways (e.g. adjust layout by removing padding and making prompt and results smaller).

I am hoping that the implementation in #959 is general enough to allow all sorts of adjustments to layout (as there are a wide range of possibilities when resizing the vim window). Once this VimResized PR is merged I would be very happy to give feedback/guidance on other PRs that allow changing the layout of a picker on the fly, e.g. toggling preview, remove all padding, etc.

---

Note: _It would be very helpful for people to test #959 to make sure that all of the edge cases have been covered ☺️_

**Review by fdschmidt93 — CHANGES_REQUESTED** [review]

**Comment by johnybx** [comment]
Any idea how to fix the prepare phase of those failing tests ?

**Comment by johnybx** [comment]
sure no problem. I will probably wait till resizing is merged and than rebase this.

The only thing which needs to be decided is how to handle case when preview is set to `false` as I mentioned [here](https://github.com/nvim-telescope/telescope.nvim/pull/1305#discussion_r726606108) - I think that we have 3 options ( or maybe I miss something ):
1. having additional setting - this is simple but does not seem to be popular option
2. initialize default previewer before calling pickers.new - this does not feel like right option because it would have to be done by every picker definition. It was suggested to do this in pickers.new but that would not work because in `opts` you get just preview setting which either previewer options like this
    ```lua
      preview = {                                                                                                                                                                                                                               
        check_mime_type = true,                                                                                                                                                                                                                 
        filesize_limit = 25,                                                                                                                                                                                                                    
        msg_bg_fillchar = "╱",                                                                                                                                                                                                                  
        timeout = 250,                                                                                                                                                                                                                          
        treesitter = true                                                                                                                                                                                                                       
      }
    ```

    or `false` if previewer is disabled. In `defaults` you get `previewer` key which is either initialized previewer or `false` if previewer is disabled.
3. while initializing config [here](https://github.com/nvim-telescope/telescope.nvim/blob/master/lua/telescope/config.lua#L727) we can save the `preview = false` to `some __preview` variable in config and set preview to default value which means that we don't expose new setting but we can still detect if preview should be hidden when picker starts. There is already [special behavior for `preview` option ](https://github.com/nvim-telescope/telescope.nvim/blob/master/lua/telescope/config.lua#L744) so we just update it.

I would probably implement 3rd option, what do you think guys ?

**Comment by l-kershaw** [comment]
I haven't gone through all of this in detail, but this implementation seems a bit overcomplicated to implement preview toggling now that we have `Picker:recalculate_layout`. This function should be able to do most of the heavy lifting in terms of managing how windows `move` with toggling the previewer, rather than having to delete windows and then create them again.

I would suggest trying to implement `Picker:hide_preview` and `Picker:unhide_preview` by copying the structure of `pickers.on_resize_window`, and then adding in the "if previewer save to hidden_previewer, etc..." type logic around that.
You should then be able to get rid of the `del_win` function for example.

---

If this is unclear at all or you disagree with me, let me know and I will do my best to clarify/discuss further 🙂

**Review by l-kershaw — COMMENTED** [review]
Had a more detailed look through this time and can only find a couple of small changes that I think are needed.

@Conni2461 is there anything else that you can think of that needs doing for this PR?

File: lua/telescope/actions/init.lua [hunk 1]
```diff
@@ -275,23 +282,14 @@ end
 actions._close = function(prompt_bufnr, keepinsert)
   action_state.get_current_history():reset()
   local picker = action_state.get_current_picker(prompt_bufnr)
-  local prompt_win = state.get_status(prompt_bufnr).prompt_win
   local original_win_id = picker.original_win_id
 
-  if picker.previewer then
-    for _, v in ipairs(picker.all_previewers) do
-      v:teardown()
-    end
```
[the remark this answers is not in the document]
Comment (johnybx) [reply lua/telescope/actions/init.lua:284]: good catch :+1: .

File: lua/telescope/pickers.lua [hunk 2]
```diff
@@ -554,9 +567,67 @@ function Picker:recalculate_layout()
...
+  elseif strategy == "descending" then
+    vim.api.nvim_win_set_cursor(win, { 1, 0 })
+    vim.api.nvim_win_set_cursor(win, { oldinfo.botline, 0 })
+    vim.api.nvim_win_set_cursor(win, oldcursor)
+  else
+    error(debug.traceback("Unknown sorting strategy: " .. (strategy or "")))
+  end
+end
+
 function Picker:hide_preview()
```
[the remark this answers is not in the document]
Comment (johnybx) [reply lua/telescope/pickers.lua]: :+1:

**Review by Conni2461 — APPROVED** [review]
Code LGTM now. Thanks :)

I havent tested this yet. I am done for today, so i wanna do that tomorrow. But @l-kershaw can also just merge this if it works and fits in his schedule (with or without this idea: https://github.com/nvim-telescope/telescope.nvim/pull/1383#pullrequestreview-794585178)

One more question. Does it make sense to map `toggle_preview` action on default? Its an awesome feature, that fixes some issue (insert number here) so it makes sense to feature it.

Again thank you for your contribution :)

**Comment by johnybx** [comment]
Default mapping is good idea :thinking:. How about `"<C-h>"` (currently not used)?

**Comment by l-kershaw** [comment]
I agree, a default mapping sounds good; however `<C-h>` doesn't seem like the right fit to me, it seems a bit arbitrary. Maybe `<M-p>`, as `p` for preview (and `<C-p>` is taken for previous item)?
I'm happy to add whichever mapping we decide on when I move the methods around.

@Conni2461 I'm happy to wait for you to test before merging; since it's a feature that people have wanted for a while, we should make sure we get it right.
Regarding your [idea](https://github.com/nvim-telescope/telescope.nvim/pull/1383#pullrequestreview-794585178) about merging into a dev branch; that definitely seems like better practice, so we should do that instead of my hacky method.
However, I'm not sure how to change the branch that a PR is wanting to merge into. I'm guessing that there is a simple way to do it that I'm just not aware of?

**Comment by tjdevries** [comment]
We can also just merge w/out default.

A decent default would be something like normal mode `za`, which is kind of like "toggle fold", which could be analagous to this and unlikely to run into any mapping problems. (see `:help za`)

I don't want to put too many things on <c-> values. Also, <c-h> is particularly bad cause that is just backspace :) you can test it out if you don't believe me (on most terminals).

**Comment by johnybx** [comment]
@l-kershaw if you create desired branch I can change target branch of pull request.

**Comment by l-kershaw** [comment]
I think that since there isn't an obvious keybind for toggling preview, lets just let users pick their own, rather than us deciding on one that a large chunk of users might not be happy with.

@johnybx I've created the `dev/layout_actions` branch, if you want to change the target of this PR to that.

@Conni2461 I'm assuming that you haven't come across any issues in your tests?

**Comment by johnybx** [comment]
@l-kershaw I updated target branch.

**Comment by l-kershaw** [comment]
Been running this PR for a few days and haven't had any issues.
Merging this into the `dev/layout_actions` branch now and will then update #1383 as previously discussed.

**Comment by l-kershaw** [comment]
Thanks for your hard work on this @johnybx 😊

**Comment by wookayin** [comment]
Great! This feature works, but when I try this I see the following error right after the preview window gets hidden:
```
E5108: Error executing lua ....vim/plugged/telescope.nvim/lua/telescope/pickers.lua:572: Cursor position outside buffer
```

**Comment by johnybx** [comment]
@wookayin I have been on this branch for long time and didn't hit this issue :thinking:. Is there some specific way how to reproduce it? Did you try with some minimal ( default ) config ?

**Comment by l-kershaw** [comment]
I think I came across something similar to @wookayin's issue when working on #1383.
Basically if you use the `"ascending"` sorting strategy, there can be times where there are less than `max_results` lines in the results buffer.
I added a check to use the actual number of lines in the buffer in the `full_layout_update` function.

@wookayin could you check to see if you can reproduce your issue when using #1383? If you're still having issues, then we can try to solve them over there 🙂

**Comment by wookayin** [comment]
@johnybx @l-kershaw  Yes, I was using `sorting_strategy = 'ascending'` with `prompt_position = 'top'`. Seems a corner case with non-default configurations :)  I can also confirm that if the default strategy (i.e., `descending`), the error does not happen. 

I also tried #1383: while doing `telescope.actions.layout.cycle_layout_next` with sorting_strategy either `ascending` or `descending` I didn't see any errors.
